Add an option to print single quotes in JSX

smirea opened the pull request:
Extending the `single-quote` option to allow using single quotes in JSX.
`single-quote` has been changed from a `bool` to a choice: `none` (false), `js` (true), `all` (new functionality, includes JSX)

Functionality [based on this comment from the long issue thread](https://github.com/prettier/prettier/issues/1080#issuecomment-378567278)

Fixes #1080.

Kovensky commented on src/language-js/printer-estree.js:
```diff
@@ -1808,7 +1808,11 @@ function printPathNoParens(path, options, print, args) {
           const value = rawText(n.value);
-          res = '"' + value.slice(1, -1).replace(/"/g, "&quot;") + '"';
+          if (options.singleQuote === "all") {
+            res = "'" + value.slice(1, -1).replace(/'/g, "&#39;") + "'";
```
Anything wrong with using `&apos;`? I'm fairly sure browsers that don't support `&apos;` can't run React either.

smirea replied:
- [ ] hmm nope, didn't know `&apos;` was a thing, will update

Kovensky replied:
It's kinda hard to find things about `&apos;`; most places talk about how it's not in the HTML4 entity list so should be avoided, but React only works on HTML5.

j-f1 commented on src/language-js/printer-estree.js:
```diff
@@ -1808,7 +1808,11 @@ function printPathNoParens(path, options, print, args) {
+          if (options.singleQuote === "all") {
+            res = "'" + value.slice(1, -1).replace(/'/g, "&#39;") + "'";
+          } else {
+            res = '"' + value.slice(1, -1).replace(/"/g, "&quot;") + '"';
```
Would it be possible for JSX strings to have the same behavior as normal strings? (i.e. `"Let’s go!"` would always use double quotes regardless of the setting, and `'Let"s go!'` would similarly always use single)

smirea replied:
hmm I think this should only apply if `singleQuote: 'all'`. The assumption would be that if you don't explicitly opt in to using single-quotes in jsx, you'll always get double-quotes, no exceptions. thoughts?

j-f1 replied:
IMO `foo="bar"` and `foo={"bar"}` should get the same quotes (except when `singleQuote` is `js` and there’s a tie between single and double quotes).

smirea replied:
that is what's happening now. For the following:

```jsx
    <div
        single='foo'
        single2={'foo'}
        double="bar"
        double2={"bar"}
        singleDouble='"'
        doubleSingle="'"
        singleEscaped={'\''}
        doubleEscaped={"\""}
        singleBothEscaped={'\'"'}
        doubleBothEscaped={"\"'"}
    />
```

it gets converted to, when `singleQuote: all`:

```jsx
    <div
        single='foo'
        single2={'foo'}
        double='bar'
        double2={'bar'}
        singleDouble='"'
        doubleSingle="'"
        singleEscaped={"'"}
        doubleEscaped={'"'}
        singleBothEscaped={'\'"'}
        doubleBothEscaped={'"\''}
    />
```

and when `singleQuote: js`:

```jsx
    <div
        single="foo"
        single2={'foo'}
        double="bar"
        double2={'bar'}
        singleDouble="&quot;"
        doubleSingle="'"
        singleEscaped={"'"}
        doubleEscaped={'"'}
        singleBothEscaped={'\'"'}
        doubleBothEscaped={'"\''}
    />
```

j-f1 replied:
How about the string `'""\''`?

j-f1 replied:
IMO it should be the same.

smirea replied:
I'm happy to change it, [however it does seem like a deliberate decision in `printString`](https://github.com/prettier/prettier/blob/master/src/common/util.js#L493) to minimize the number of escapes by counting the types of quotes in each, and effectively disregarding the `singleQuote` option. In other words:

`'""\''` - will always be single-quote, regardless of options, since it has 2 double and 1 single inside
`'"\'\''` - will always be double-quote, regardless of options, since it has 1 double and 2 single inside

In order to force this to be "always the same" I'd basically have to bypass or remove that logic. 
I have no qualms doing so, but just wanted to double-check that that's the desired approach

j-f1 replied:
Sorry if I’m not communicating clearly :/

What I meant by “always be the same” is to use the `printString` algorithm for JSX, too.

smirea replied:
oooh, ok, yeah sorry I thought you meant the other way around, makes sense now, will do

Kovensky replied:
I agree, if escapes should be needed, unless it's like only one, it's much better to just convert it to a string expression instead of keeping it as a raw string with a bunch of HTML escapes. The HTML escapes also only make sense for ReactDOM; they likely won't do the right thing on react native.

smirea replied:
I pushed an update that uses the JS print function if any escape is needed, bypassing the "convert to html entities" escape method altogether. 

----

I did end up going into a few rabbit holes trying to get `foo="bar"` and `foo={"bar"}` to sync up which I might need to some help on [mentioned above](https://github.com/prettier/prettier/pull/4798#discussion_r199832547). 
The core issue is that `print` needs to now behave differently depending on what context it's in. i.e. `printString` needs to:
- use single quotes when singleQuote: all
- use single quotes when singleQuote: js **and** it's not a JSX attribute
- use double quotes when singleQuote: js **and** it is a JSX attribute 

The problem I'm having with the above is that the expression inside `{"bar"}` can be any JS expression, that sometimes will be a string literal, sometimes a concatenation of string literals and most times any other random expression. I tried going down the path of passing an extra contextual option to `print` when it's called from the JSX context, but that just ends up converting all strings to double-quotes inside the entire expression

TL; DR: when singleQuote: js, I cannot see how to sync `foo="bar"` with `foo={"bar"}` without also having things like `foo={() => { const a = "bar"; }}` also use double-quotes.

lydell commented on src/language-js/printer-estree.js:
```diff
@@ -1808,7 +1808,21 @@ function printPathNoParens(path, options, print, args) {
+          const hasDouble = innerValue.includes('"');
+
+          if (options.singleQuote === "all") {
+            if (hasSingle && hasDouble) {
```
I think we should count the number of singles and doubles and choose surrounding quotes so that the number of escapes is minimized, like we do for JS strings.

But wait a minute – can there ever be both `'`s and `"`s at the same time in an attribute value? I don't think so. And is there are risk of getting all `'`s and `"`s turned into `&apos;` and `&quot;` as you run Prettier in between edits? Should we unescape unnecessary `&apos;` and `&quot;`?

smirea replied:
valid point, I just realized as well that it's not valid to escape quotes in pure JSX attributes, i.e.

this is invalid: `<div id="invalid \" jsx" />`
this works though: `<div id={"valid \" jsx"} />`

I'm updating the format to use the same logic from `printString`, [there's a previous thread going through this discussion in more depth above](https://github.com/prettier/prettier/pull/4798#discussion_r199756260)

Kovensky replied:
It's not invalid; it just won't do what you mean 😄

It's better to just convert to string expressions if you need any escaping.

rattrayalex commented:
@suchipi thoughts on putting this to a private vote amongst the maintainers or something? Perhaps over email? 

We're all in agreement that we hate bikeshedding discussions, but it's not clear that there's overwhelming evidence one way or another on this one (eg; community demand seems strong, including from some luminaries). Eg; prettier's existence/stance doesn't seem to prevent teams from having conversations about this, it's just "do we use prettier when it lacks this feature".

Kovensky commented:
I can say that I haven't yet adopted prettier _specifically_ because of its lack of support for single quote JSX attributes. We don't write conversational english in our raw attributes, there is no need to care about the need to handle a potential apostrophe in your text, and forcing the attributes to be double quotes when all your other strings are single quotes is disorienting.

But I'm not an influencer I guess.

All this discussion was already had in the long issue thread that OP mentioned.

lydell commented on src/language-js/printer-estree.js:
```diff
@@ -1807,8 +1807,24 @@ function printPathNoParens(path, options, print, args) {
+
+          // If we need to escape quotes, just use the JS handler
+          if (hasSingle && hasDouble) {
+            res = path.call(print, "value");
```
1. Does this mean that we would end up in the `printString` function in `src/common/util.js`? That would not be OK, because StringLiterals used as JSX attribute values do not have the same escaping rules as StringLiterals in other places.

2. Can this ever happen? Is it better to `throw` here?

smirea replied:
2. I'm not sure about this. I was under the impression that you can't escape quotes inside JSX string literal attributes, but the previous code handled escaping and there is also a comment above suggesting it might be possible. 

1. If escaping is not possible, then this can just be removed and not even worry about escaping. If it is, do you have some guidance on this? My first thought would be to just convert it into an actual expression and then use the underlying `printString` (from the above discussion we landed on using printString if an escape is needed to keep behavior parity)

lydell replied:
HTML escapes are used instead of backslashes.

smirea replied:
I update the above to make it a JSXExpressionContainer if an escape is needed so the `printString` function will properly escape now

rattrayalex commented:
Yeah @Kovensky let's please avoid rehashing conversations 😄, it's clear from #1080 there's strong community demand, and countless reasons have been mentioned in that thread. 

It's a clear tradeoff between "give the people what they want" and "never add options", which is just up to the maintainers. 

In any case @smirea thanks for the work on this PR – the better it looks, the more likely it'll be accepted, I hope 🙂

lydell commented on src/language-js/printer-estree.js:
```diff
@@ -1807,8 +1807,24 @@ function printPathNoParens(path, options, print, args) {
+
+          // If we need to escape quotes, just use the JS handler
+          if (hasSingle && hasDouble) {
+            res = concat(["{", path.call(print, "value"), "}"]);
```
Unfortunately, this isn’t correct either. `<e attr="&nbsp;" />` is not equivalent to `<e attr={"&nbsp;"} />`. And I don’t think we should try converting HTML escapes into JS escapes. Again, I don’t think it is possible to get into this situation when the AST comes from a parser.

smirea replied:
ok I'm not experienced with all the parsing quirks and [I'm getting conflicting replies](https://github.com/prettier/prettier/pull/4798#discussion_r199989155) on how to handle this.

From your comments it seems that it's not feasible to use the `printString` method itself to do the work, but we want to get the same functionality as `printString` which lease the option of abstracting the preferred code selection logic 

[I opened a 2nd branch with that change](https://github.com/smirea/prettier/compare/feat-option-singleQuote-jsx...smirea:feat-option-singleQuote-jsx-escape?expand=1) can I get your thoughts on it before updating this one? I'm trying to get a grasp on this before spamming everyone else with more updates

j-f1 replied:
Looks great! The only change I would make is to remove unnecessary escapes (i.e. change `'&apos;'` to `"'"` instead of `"&apos;"`), but I’m not at a computer to check if your branch does that already.

lydell replied:
That second branch looks good.

If you’re interested, here’s some stuff to read on the quirkyness of escaping in JSX:

https://github.com/yannickcr/eslint-plugin-react/issues/1449
https://github.com/yannickcr/eslint-plugin-react/pull/1458

smirea replied:
@j-f1 it did not, it does now
@lydell thanks for the feadback + quirkiness docs, it was an interesting read!

smirea commented:
hi all, bump on this. I think I've addressed all the pending questions in the last commit from a few weeks ago. 

Can y'all take another look so I can do a final rebase

suchipi commented:
I still don't think that we should add this, but if we do, for what it's worth, some plugins are relying on singleQuote being boolean; I'm not sure if they can use support info to figure out how they're supposed to handle that.

ikatyang commented:
We should change options to be language-specific, e.g. `"javascript/singleQuote": "something"`, though not sure how should it work.

smirea commented:
this feature is backwards compatible with the old signature via the config.
right now `singleQuote: true` redirects to `singleQuote: js` and `false` converts to `none` with an accompanying deprecation warning, so old modules will still work.

We could support all 5 options (true/js false/none all) without a deprecation to not have it be slightly awkward for non-js projects, my 2 cents is that it's better to bite the bullet sooner rather than later

Example of the current behavior:

<img width="745" alt="screen shot 2018-07-24 at 9 48 02 am" src="https://user-images.githubusercontent.com/544097/43142553-afeaf296-8f26-11e8-9478-e425f5dfb21c.png">

suchipi commented:
@smirea I mean like here: https://github.com/prettier/plugin-lua/blob/7e41cece6200de35244e1bea118452cb5efad5fa/src/printer.js#L570

Any prettier plugin could be relying on that option being a boolean.

bovas85 commented:
Any luck on this to be merged if approved?

rattrayalex commented:
Quick 👍 / 👎  – maintainers only please – @j-f1 @suchipi @lydell @ikatyang @vjeux @jlongster @azz – should we merge this?

ikatyang commented:
I have no opinion on this PR but it shouldn't break other plugins https://github.com/prettier/prettier/pull/4798#issuecomment-407581046.

rattrayalex commented:
@ikatyang - agreed that that's a blocker to merge, though I'm guessing the author of this PR would rather not deal with implementing a solution for that until it's clear we want the change.

smirea commented:
I'm happy to fix the blockers if there is a desire to merge this, there's still a significant community desire for this feature :)

bovas85 commented:
I think this is a very much needed feature as it blocks everyone using standard.js along with prettier

j-f1 commented:
Quick vote for @prettier/maintainers (only maintainers please): 👍 if this should be merged, 👎 if it shouldn’t.

lydell commented:
What about `always`, `never`, `except-jsx` as the option values?

j-f1 commented:
How about `false`, `true`, and @lydell’s `"except-jsx"`? This would mean that falsy values mean double quotes and truthy values single quotes, and the JS parser would look more closely at the exact values.

ikatyang commented:
FYI, we could forward singleQuote to jsSingleQuote to avoid breaking change, see #5020 for more info.

j-f1 commented:
I’ll check the votes on Monday and see what the general opinion is.

j-f1 commented:
Looks like there are no 👎 votes. Can you fix the conflicts @smirea?

smirea commented:
perfect! I'll take a look and push the updates over the weekend, looking forward to playing with @ikatyang's new argument forwarding

sheerun commented:
Finally :) This rule and `space-before-function-paren` (I hope it'll be implemented as well) are last that allow to format according to standard.js rules without using prettier-eslint

smirea commented:
reviving this with updates, added `jsSingleQuote` and forwarding it to `singleQuote` as per @ikatyang 's suggestion so now this is fully backwards compatible. 🎉 

Can I get some final feedback on this and what would be the ideal format before I finalize tests and update snapshots?

ikatyang commented:
After second thought, it actually can be an independent bool option (`--jsx-single-quote`), there's no need to forward them (unnecessary complexity). What do you think?

smirea commented:
Personally personally, I would've not even added an extra option and have jsx follow js quotes and not have the notion of "we use a type of quote everywhere in js, except for this place" ... but I get how that is not possible now

In lieu of that, I think the `--jsx-single-quote` actually makes the most sense, it did feel a bit weird having an option override another

I'll give it a day and if there aren't any other suggestions i'll go ahead and add `--jsx-single-quote` and wrap it up

j-f1 commented:
An idea I had recently: (this may be completely crazy)

- `singleQuote: false`: never output single quotes (except in e.g. `'"'`)
- `singleQuote: true`: always output single quotes (except in e.g. `"'"`)
- `singleQuote: ['js']`: single quotes only in JS
- `singleQuote: ['jsx', 'foo', 'bar']`: single quotes in JSX but not JS, plugins can make use of the `foo` and `bar` values.

This preserves backward compatibility since `singleQuote: true` matches everything. It also allows simple plugins to check the truthiness of the `singleQuote` option and since we check `singleQuote.includes('jsx')`, other plugins can add their own customizations if needed.

---

Is this excessively complex?

smirea commented:
Ok I've updated the PR with:

- use `jsx-single-quote` plain ol' boolean, removed all the `vnopts` magic
- updated docs
- updated snapshots

Can y'all have another gander and see if this checks out?

lydell commented on docs/options.md:
```diff
@@ -60,7 +60,7 @@ Use single quotes instead of double quotes.
 Notes:
 
-- Quotes in JSX will always be double and ignore this setting.
+- JSX Quotes will ignore this setting. see `jsx-single-quote`
```
```suggestion
- JSX quotes ignore this option – see [jsx-single-quote].
```

smirea replied:
will update

lydell replied:
I think this new suggestions thing would have let you just press a button here on GitHub to apply the change..

smirea replied:
i've never used it before, but it seemed to want to do a commit for every applied suggestion when I tried, which seemed a bit excessive

lydell replied:
We always use “squash and merge” anyway so it doesn’t matter

lydell commented on docs/options.md:
```diff
@@ -69,6 +69,14 @@ See the [strings rationale](rationale.md#strings) for more information.
 
+## JSX Quotes
+
+Use single quotes in JSX
```
```suggestion
 Use single quotes instead of double quotes in JSX.
```

smirea replied:
will update

lydell commented on docs/options.md:
```diff
@@ -69,6 +69,14 @@ See the [strings rationale](rationale.md#strings) for more information.
+
+| Default | CLI Override | API Override   |
+| ------- | ------------ | -------------- |
+| `false`  | `--jsx-single-quote`  | `jsxSingleQuote: <bool>` |
```
I think you need to run Prettier on this file to make `yarn lint-docs` pass :)

smirea replied:
I had no idea that was a thing, it's not part of the [Contributing docs](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md) :trollface: . Happy `yarn lint-docs --fix` works like a charm 👍

lydell commented on tests/markdown/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -595,7 +595,7 @@ Default | CLI Override | API Override
 
 Notes:
-* Quotes in JSX will always be double and ignore this setting.
+* JSX will always be double quotes and ignore this setting. To use single quotes in JSX use \`jsxSingleQuote\`
```
Not sure if this snapshot really should have been updated.

smirea replied:
I'm not sure either, will revert

lydell commented on tests/quotes/jsfmt.spec.js:
```diff
@@ -1,2 +1,2 @@
 run_spec(__dirname, ["flow"]);
-run_spec(__dirname, ["flow"], { singleQuote: true });
+run_spec(__dirname, ["flow"], { jsxSingleQuote: true });
```
Why was this changed?

smirea replied:
ah, artifact from the old functionality. will revert

smirea commented:
hey all, all comments were addressed and everything is done from my end in the last commit. 
Any final thoughts? If not, could we look into actually merging this finally? 😄

lydell commented on src/common/util.js:
```diff
@@ -430,43 +431,57 @@ function getIndentSize(value, tabWidth) {
 
-function printString(raw, options, isDirectiveLiteral) {
+function getPreferredQuote(raw, preferredQuote) {
+  assert(
```
Looks like we don’t use `assert` anywhere else. Maybe we can just remove this `assert` (no need to validate I think)

smirea replied:
removed

ikatyang commented:
@j-f1 since you requested changes before, would you like to do a final review?

j-f1 commented:
Just needs conflicts resolved.

smirea commented:
conflicts fixed, snapshots updated, checks passed, we're green across the board!

j-f1 commented:
Thank you for contributing to Prettier!
